dockstore-integration-testing/src/test/java/io/dockstore/client/cli/WorkflowIT.java

```diff
@@ -1092,6 +1090,43 @@ public void testGettingImagesFromQuay() {
         verifyTRSImageConversion(versions, "1.0", 3);
     }
 
+    /**
+     * Tests that snapshotting a workflow version fails if any of the images have no tag, use the 'latest' tag, or are specified using a parameter.
+     */
+    @Test
+    public void testSnapshotImageFailures() {
+        final ApiClient webClient = getWebClient(USER_2_USERNAME, testingPostgres);
+        WorkflowsApi workflowsApi = new WorkflowsApi(webClient);
+        Workflow workflow = manualRegisterAndPublish(workflowsApi, "dockstore-testing/hello-wdl-workflow", "", DescriptorType.WDL.toString(), SourceControl.GITHUB, "/Dockstore.wdl", false);
+        WorkflowVersion noTagImageVersion = workflow.getWorkflowVersions().stream().filter(v -> v.getName().equals("noTagImage")).findFirst().get();
+        WorkflowVersion latestTagImageVersion = workflow.getWorkflowVersions().stream().filter(v -> v.getName().equals("latestTagImage")).findFirst().get();
+        WorkflowVersion parameterImageVersion = workflow.getWorkflowVersions().stream().filter(v -> v.getName().equals("parameterImage")).findFirst().get();
+
+        // Test that the snapshot fails for a workflow version containing an image with no tag
+        try {
+            noTagImageVersion.setFrozen(true);
+            workflowsApi.updateWorkflowVersion(workflow.getId(), Collections.singletonList(noTagImageVersion));
+        } catch (ApiException ex) {
+            Assert.assertTrue(ex.getMessage().contains("Snapshot for workflow version noTagImage failed because not all images are specified using a digest or a valid tag."));
+        }
+
+        // Test that the snapshot fails for a workflow version containing an image with the 'latest' tag
+        try {
+            latestTagImageVersion.setFrozen(true);
+            workflowsApi.updateWorkflowVersion(workflow.getId(), Collections.singletonList(latestTagImageVersion));
+        } catch (ApiException ex) {
+            Assert.assertTrue(ex.getMessage().contains("Snapshot for workflow version latestTagImage failed because not all images are specified using a digest or a valid tag."));
+        }
+
+        // Test that the snapshot fails for a workflow version containing an image specified using a parameter
+        try {
+            parameterImageVersion.setFrozen(true);
+            workflowsApi.updateWorkflowVersion(workflow.getId(), Collections.singletonList(parameterImageVersion));
```

Comment:
You need an assertFail() at the end of the try blocks or else they may all silently pass (never triggering catch).